feat: Add download started and download completed callbacks

atlanticaccent opened the pull request:
<!--
Update "[ ]" to "[x]" to check a box

Please make sure to read the Pull Request Guidelines: https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

### What kind of change does this PR introduce?
<!-- Check at least one. If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers. -->

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
<!-- If yes, please describe the impact and migration path for existing applications in an attached issue. -->

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### Other information
Windows works completely - the callbacks fire and the file downloads. The download location can even be changed through the crude API I've implemented (a `&mut String` passed to the download started callback). One note on changing download location paths is that Universal Naming Convention file paths do not work (they start with slashes and a question mark) - these are generated when calling `canonicalize` so an alternative solution is necessary.

Linux seems to be working to some extent - both callbacks fire, however controlling where the file is saved seems to have no effect and causes the download to stall indefinitely. Not setting the download path causes bizarre behaviour where webkitgtk either claims that the download has succeeded (to where is unclear) or that the path (which is full of invalid characters) is occupied.

~~I attempted a macOS implementation however seem to have failed completely. I have a `DownloadDelegate` being created and set, and a couple of what I think are necessary methods implemented on the `WKWebview`, but nothing seems to get the actual delegate functions to fire.~~
The macOS implementation is effectively on par with the linux implementation, in that all the callbacks now fire correctly, but setting the download location causes the entire download to fail, and not setting it does not seem to actually write out a download/file.

atlanticaccent committed:
Add download handler to webview attributes

atlanticaccent committed:
Potentially implement download event for windows

atlanticaccent committed:
Add download event example

atlanticaccent committed:
Implement download event for webkitgtk

atlanticaccent committed:
Attempt to write example writing to tempdir

atlanticaccent committed:
Add download compete handler, fix example

atlanticaccent committed:
Update doc

atlanticaccent committed:
Fix webkitgtk implementation for download handlers

Commit:
Attempt to implement download events on macOS

atlanticaccent committed:
Use more reliable URLs

atlanticaccent committed:
Improve gtk implementation

atlanticaccent committed:
Add more details to example

Commit:
Attempt to write tempfile to documents

Commit:
Fix download delegate funcs to implement on navdelegate on macOS

Also adds the complete and failed callbacks.

atlanticaccent committed:
dummy commit

atlanticaccent committed:
Update webkit2gtk

amrbashir commented on src/webview/mod.rs:
```diff
@@ -126,6 +126,19 @@ pub struct WebViewAttributes {
+  pub download_handlers: Option<(
+    Box<dyn FnMut(String, &mut String) -> bool>,
+    Box<dyn Fn() -> Box<dyn Fn(String, bool) + 'static> + 'static>
+  )>,
```
Should be split into two handlers, `download_handler` and `download_completed_callback`

amrbashir commented on src/webview/webkitgtk/web_context.rs:
```diff
@@ -198,6 +207,16 @@ impl WebContextExt for super::WebContext {
+      Box<dyn Fn() -> Box<dyn Fn(String, bool) + 'static> + 'static>,
+    ),
+  ) {
+    actually_register_download_handler(self, handler);
```
just inline the function here, no need to extract the logic in a separate function

amrbashir commented on src/webview/mod.rs:
```diff
@@ -126,6 +126,19 @@ pub struct WebViewAttributes {
+  /// saved and a `bool` indicating if the download succeeded.
+  pub download_handlers: Option<(
+    Box<dyn FnMut(String, &mut String) -> bool>,
+    Box<dyn Fn() -> Box<dyn Fn(String, bool) + 'static> + 'static>
```
why does this need to be a function that returns a function?

atlanticaccent replied:
I'm probably not going to explain this well, but I'll give it a shot:

If I just use a regular function, we run into the issue that (I think) we need a mutable copy of the download completed callback _every time the download started function is called_. To do that with a "raw" closure would mean cloning the closure, but I can't find a way to define the closure as cloneable when storing it - `Fn` closures are cloneable if all their captured variables are cloneable, but I can't see a way of specifying that.

An alternative solution (again, I think) would be to wrap the "raw" closure in an `Rc`, which would make it as cloneable as we like.

I can't actually tell if this is the same stories on other platforms, but the fact it's the case on windows basically makes it unavoidable.

amrbashir replied:
Yeah `Rc` should work, it is better than having this level of indirection. cc @wusyong

atlanticaccent replied:
Alternatively this https://stackoverflow.com/a/65204236 may be a viable solution, though it means implementing a relatively complex custom trait that may make the API more opaque to users

atlanticaccent replied:
I'll use the `Rc` solution as it seems the simplest

amrbashir commented on src/webview/mod.rs:
```diff
@@ -126,6 +126,19 @@ pub struct WebViewAttributes {
+  /// The second closure is fired when the download completes, with a `String` representing the path to where the download was
+  /// saved and a `bool` indicating if the download succeeded.
+  pub download_handlers: Option<(
+    Box<dyn FnMut(String, &mut String) -> bool>,
```
Probably better to use the `Path` type
```suggestion
    Box<dyn FnMut(String, &mut Path) -> bool>,
```

atlanticaccent replied:
Switched to using `&mut PathBuf` - `Path` size can't be known at compile time so we can't take a mutable reference.

amrbashir replied:
Yeah that's fine too.

atlanticaccent committed:
Match changes on dev

atlanticaccent committed:
Split download handlers

Now has handler for download start/deny and download completed callback

atlanticaccent committed:
Propagate split of download handlers to win impl

atlanticaccent committed:
Switch to mutable ref PathBuf instead of string

atlanticaccent committed:
Wrap download_completed callback in Rc

This avoids the indirection of the closure builder pattern whilst still
solving the lifetime issues.

keiya01 commented:
What is this PR status?

atlanticaccent committed:
Merge branch 'dev' into implement-download-event

atlanticaccent committed:
Windows formatting

atlanticaccent committed:
Fix merge in linux implementation

Downloads still don't actually complete though

lucasfernog commented:
What's the status here?

atlanticaccent commented:
I'm attempting to get the work I had done fixed up so it's up to date with master, however the outstanding issues in linux (and I assume macOS) still remain. Specifically, downloads don't complete and modifying the target file path makes no difference (or causes the download to fail I haven't narrowed it down).
I may be able to get the macOS version up to date soon, but macOS still isn't my primary environment

atlanticaccent committed:
Fix macOS implementation + refactor

atlanticaccent committed:
Rework example

Now holds temp_dir as long as necessary to prevent premature drop (and
thus deleting our example file). Requires the use of the Rc<RefCell>
pattern for interior mutability across a shared reference - this
(should) be safe, as the entire example is single threaded, and any
mutable borrows should be dropped by the time a borrow ocurrs.

atlanticaccent committed:
Merge branch 'dev' into implement-download-event

atlanticaccent committed:
Ignore unused parameter

atlanticaccent committed:
Improve download example handling of empty path

atlanticaccent committed:
Formatting

atlanticaccent committed:
Attempt to improve linux behaviour

atlanticaccent commented:
The macOS implementation is now fixed, and the entire PR should be up to date with master (barring any mistakes made while merging). I have also managed to fix the macOS issue where downloads would spuriously fail. I still don't have a proper linux setup though, so can't test that platform fully.

There's a couple things that still need done. Completed handlers still need to be completely decoupled from started handlers on all platforms so that they fire regardless of whether a started handler is present. I also think that the completed handler should have the URI of the downloaded resource added as a parameter, as otherwise there is no reliable way to tell what download the handler is actually being fired for.

atlanticaccent commented:
@lucasfernog @amrbashir

amrbashir commented:
> I still don't have a proper linux setup though, so can't test that platform fully.

don't worry, we can test it later once you're finished with the PR

> I also think that the completed handler should have the URI of the downloaded resource

yeah that would be a good addition if it is possible

atlanticaccent commented:
> > I also think that the completed handler should have the URI of the downloaded resource

> yeah that would be a good addition if it is possible

I'm actually thinking that it might even be worth generating some kind of unique identifier to attach to downloads (like a GUID) as URIs are not actually a completely unique identifier for download operations - say, for example, two downloads of the same resource are triggered in quick succession. In the example, it's also not sufficient to assume the first download to complete was the download that was started first, as it may have stalled for any number of reasons whilst the download initiated second completed first. An alternative to GUIDs would be something like attaching the system time the download was initiated at.

amrbashir commented:
GUIDs is a bit overkill. I think these download handlers could be used to spawn a dialog for the user to choose the file path and so it should be the download handler responsibility to avoid conflicts

atlanticaccent committed:
Attempt to fix Windows compile errors

atlanticaccent committed:
Separate download complete handler from download started handler

atlanticaccent committed:
Formatting

atlanticaccent committed:
Take closure by mutable reference (windows)

atlanticaccent committed:
Workaround mutable borrow issues on windows

atlanticaccent committed:
Separate download started handler from finished on linux

atlanticaccent committed:
Potentially improve setting output path on linux

atlanticaccent committed:
Add original download's url as parameter to download completed handler

atlanticaccent committed:
Formatting

atlanticaccent committed:
Standardise terminology (replace `callback` with `handler`)

amrbashir commented on src/webview/mod.rs:
```diff
@@ -160,6 +160,31 @@ pub struct WebViewAttributes {
+  ///
+  /// ## Platform-specific
+  ///
+  /// - **Windows:** The assigned path cannot include a UNC (Universal Naming Convention) prefix. This prefix is
+  /// sometimes included when canonicalizing paths in order to differentiate between networked filesystems.
```
We can just use `dunce` crate to strip UNC from the path after it has been modified by the handler

atlanticaccent committed:
Use dunce to attempt to remove UNC prefixes on windows

atlanticaccent committed:
Fix incorrect function signature on macOS

atlanticaccent committed:
Improve docs

atlanticaccent committed:
Enable devtools in example

atlanticaccent committed:
Include blob download example

macOS implementation works, but may rely on an incorrect assumption -
clicking the download link seems to return a random blob URL, so we
just grab the first saved blob URL

atlanticaccent committed:
Formatting

atlanticaccent commented:
@amrbashir I believe I've addressed your comments, plus cleaned some things up. Would you be able to give it another look over?

amrbashir commented on src/webview/mod.rs:
```diff
@@ -160,6 +160,26 @@ pub struct WebViewAttributes {
+  ///
+  /// - **macOS**: The second parameter indicating the path the file was saved to is always empty, due to API
+  /// limitations.
+  pub download_completed_handler: Option<Rc<dyn Fn(String, String, bool) + 'static>>,
```
could we change this also to provide `PathBuf`
```suggestion
  pub download_completed_handler: Option<Rc<dyn Fn(String, PathBuf, bool) + 'static>>,
```

amrbashir commented on src/webview/mod.rs:
```diff
@@ -416,6 +438,38 @@ impl<'a> WebViewBuilder<'a> {
+  /// limitations.
+  pub fn with_download_completed_handler(
+    mut self,
+    download_completed_handler: impl Fn(String, String, bool) + 'static,
```
ditto
```suggestion
    download_completed_handler: impl Fn(String, PathBuf, bool) + 'static,
```

amrbashir reviewed (changes requested):
Other than these two nit-picks, the Linux/Windows implementation looks good, as for macOS, I will leave it for @wusyong or @lucasfernog to approve.

atlanticaccent committed:
Address comment regarding passing pathbuf to completion handler

atlanticaccent committed:
Separate download completed from started handler on macOS

amrbashir reviewed (dismissed):
LGTM now for Windows and Linux

atlanticaccent committed:
Formatting

amrbashir commented on src/webview/wkwebview/mod.rs:
```diff
@@ -329,66 +336,240 @@ impl InnerWebView {
+      {
+        let cls = match ClassDecl::new("UIViewController", class!(NSObject)) {
+          Some(mut cls) => {
+            cls.add_ivar::<*mut c_void>("function");
+            cls.add_ivar::<*mut c_void>("HasDownloadHandler");
```
Can we extract these two strings into some consts so it is easier to track/change in the future? Actually any ivar used more than one time should be in a const

wusyong replied:
There are many more ivar like this. I think we should do this in another PR.

amrbashir reviewed (changes requested):
Sorry for taking so long to merge this PR, our macOS expert is currently on vacation

wusyong committed:
Merge branch 'dev' into implement-download-event

wusyong committed:
Move download ffi function to download module

wusyong committed:
Add change file

wusyong committed:
Correct the name of download start method

wusyong commented:
Thanks for this amazing PR @atlanticaccent !
